File: GeeksforGeeks/Python3/Medium/Maximum Product of Increasing Subsequence of Size 3/maximum-product-of-increasing-subsequence-of-size-3.py
#User function Template for python3

import sys
import bisect

class Solution:
    def countArray (self, arr, n) : 
        # Complete the function
        g = [-1] * n
        maxi = arr[-1]
        for i in range(n - 2, -1, -1):
            g[i] = max(maxi, arr[i])
            maxi = g[i]
        s = [arr[0]]
        re = -sys.maxsize
        l = [-1]
        for i in range(1, n - 1):
            x = bisect.bisect_left(s, arr[i])
            if x != 0 and g[i] > arr[i] and arr[i] * g[i] * s[x - 1] > re:
                re = arr[i] * g[i] * s[x - 1]
                l = [s[x - 1], arr[i], g[i]]
            bisect.insort(s, arr[i])
        return l

#{ 
 # Driver Code Starts
#Initial Template for Python 3


for _ in range(0,int(input())):
    n = int(input())
    arr = list(map(int, input().strip().split()))
    ans = Solution().countArray(arr, n)
    if(ans[0] == -1):
        print("Not Present")
    else:
        print(*ans)
# } Driver Code Ends

File: GeeksforGeeks/C++/Medium/Maximum Product of Increasing Subsequence of Size 3/maximum-product-of-increasing-subsequence-of-size-3.cpp
//{ Driver Code Starts
#include<bits/stdc++.h> 
using namespace std; 


// } Driver Code Ends
// Function to find the max product subsequence of size 3
// arr: given input array
// n: size of array
// res: array which contains the output
void maxProductSubsequence(int *arr , int n, int *res)
{
	// Your code here
	if(n <= 2){
	    res[0] = res[1] = res[2] = 0;
	    return;
	}
	long long int ans = 0;
	vector<int>rightMax(n, 0);
	rightMax[n-1] = arr[n-1];
	for(int i = n-2; i>=0; i--)rightMax[i] = max(arr[i], rightMax[i+1]);
	multiset<int>s;
	s.insert(arr[0]);
    for(int i = 1; i < n-1; i++){
        s.insert(arr[i]);
        auto it = s.lower_bound(arr[i]);
        if(it != s.begin() and rightMax[i+1] > arr[i]){
            --it;
            long long int val = *it * 1LL * arr[i] * rightMax[i+1];
            if(val > ans){
                res[0] = *it, res[1] = arr[i], res[2] = rightMax[i+1];
                ans = val;
            }
        }
    }
}


//{ Driver Code Starts.

// Driver Program 
int main()
{ 
    // ios_base::sync_with_stdio(false);
    // cin.tie(NULL);
	
	int T;
	cin>>T;
	while(T--){
	    int num;
	    cin>>num;
	    int arr[num];
	    for(int i = 0; i<num; ++i)
	        cin>>arr[i];
	    
	    int res[3] = {0};
	    maxProductSubsequence(arr, num, res);
	    
	    if(res[0] == 0)
	        cout<<"-1" << endl;
	    else{
    	    for(int i = 0; i<3; ++i)
    	        cout<<res[i]<<" ";
    	    cout << endl;
	    }
	}
} 


// } Driver Code Ends